### PR TITLE
Tighten bit shifts

### LINKED ISSUE #827: Define behavior of negative shifts
The behavior of a shift operation when the right operand indicating the amount to shift is negative is somewhat vague in many languages. This issue is of concern regardless of whether the shift is arithmetic or logical.

Options as I see them include:
1. Reverse the direction of the shift
2. Treat all right operands as unsigned
3. produce an error negative shifts

1. might give unexpected results and introduces the weird situation where a left shift might become a right shift, which introduces the arithmetic or logical shift issue. If arithmetic right shifts are limited to signed values, as might end up being the case, but left shifts are open to unsigned, this could be a back door method of applying an arithmetic shift to an unsigned value.
2. eliminates the negative shift problem, but whenever a negative value is encountered, we'll be getting an overshift, which just passes the buck into another issue.
3. perfectly fine to produce a compile error when the right side is a literal, but if it's a variable, the error has to be runtime, which is often not very convenient in shading languages.

### LINKED ISSUE #828: Define behavior of too-large shift values
How do we treat `1 << 32`?

### PR DESCRIPTION
**Pull request by dneto0**
- RHS of shift must be unsigned type
- number of bits to shift is the RHS modulo the component width
  of the LHS

Fixes #827
Fixes #828

### CODE REVIEW

File: wgsl/index.bs [hunk 1]
```diff
@@ -2008,32 +2008,42 @@ Issue: (dneto): Bitwise-complement is under discussion.  https://github.com/gpuw
...
-          *e2* : i32<br>
+          *e2* : u32<br>
        <td class="nowrap">`e1 >> e2` : i32
-       <td>Arithmetic shift *e1* right by *e2* bits, i.e. replicating the sign bit of *e1* at most significant bits (OpShiftRightArithmetic)
-  <tr><td>*e1* : *T*<br>
-          *e2* : *T*<br>
-          *T* is vec*N*&lt;i32&gt;
-       <td class="nowrap">`e1 >> e2` : *T*
+       <td>Arithmetic shift right:<br>
+           Shift *e1* right, copying the sign bit of *e1* into the most significant positions,
```
Comment (kvark) [inline wgsl/index.bs:2040]: I wonder if this should be described as an algorithm in bikeshed

File: wgsl/index.bs [hunk 2]
```diff
@@ -2008,32 +2008,42 @@ Issue: (dneto): Bitwise-complement is under discussion.  https://github.com/gpuw
...
   <tr><td>*e1* : i32<br>
-          *e2* : i32<br>
+          *e2* : u32<br>
        <td class="nowrap">`e1 >> e2` : i32
-       <td>Arithmetic shift *e1* right by *e2* bits, i.e. replicating the sign bit of *e1* at most significant bits (OpShiftRightArithmetic)
-  <tr><td>*e1* : *T*<br>
-          *e2* : *T*<br>
-          *T* is vec*N*&lt;i32&gt;
-       <td class="nowrap">`e1 >> e2` : *T*
+       <td>Arithmetic shift right:<br>
```
Comment (kvark) [inline wgsl/index.bs:2039]: is there a value in calling it "arithmetic shift" versus something simpler, like a "signed shift"?

Comment (dneto0) [reply wgsl/index.bs:2039]: "Arithmetic shift" is a common name for it.

https://en.wikipedia.org/wiki/Arithmetic_shift

File: wgsl/index.bs (hunk 1 above)
Comment (dneto0) [reply wgsl/index.bs]: I didn't know about that feature.  It's better that way.  See the next commit.

File: wgsl/index.bs [hunk 3]
```diff
@@ -2007,33 +2007,43 @@ Issue: (dneto): Bitwise-complement is under discussion.  https://github.com/gpuw
...
-          *T* is vec*N*&lt;i32&gt;
-       <td class="nowrap">`e1 >> e2` : *T*
+  <tr algorithm="scalar arithmetic shift right"><td>|e1| : i32<br>
+          |e2| : u32<br>
+       <td class="nowrap">|e1| `>>` |e2| : i32
+       <td>Arithmetic shift right:<br>
+           Shift |e1| right, copying the sign bit of |e1| into the most significant positions,
+           and discarding the least significant bits.
+           The number of bits to shift is the value of |e2| modulo the bit width of |e1|.
+           (OpShiftRightLogical)
```
Comment (dj2) [inline wgsl/index.bs:2043]: OpShiftRightArithmetic?

File: wgsl/index.bs [hunk 4]
```diff
@@ -2007,33 +2007,43 @@ Issue: (dneto): Bitwise-complement is under discussion.  https://github.com/gpuw
...
+       <td>Shift left:<br>
+           Shift |e1| left, inserting zero bits at the least significant positions,
+           and discarding the most significant bits.
+           The number of bits to shift is the value of |e2| modulo the bit width of |e1|.
+           (OpShiftLeftLogical)
+  <tr algorithm="vector shift left"><td>|e1| : vec|N|&lt;|T|&gt;<br>
+          |e2| : vec|N|&lt;u32&gt;<br>
+          |T| is *Int*
+       <td class="nowrap">|e1| `<<` |e2| : vec|N|&lt;|T|&gt;
        <td>Component-wise shift left (OpShiftLeftLogical)
```
Comment (dj2) [inline wgsl/index.bs:2023]: Should we say something like a[i] << b[i] ?

File: wgsl/index.bs (hunk 3 above)
Comment (dneto0) [reply wgsl/index.bs:2043]: ugh, copypasta. Right thing to do is to delete these.

File: wgsl/index.bs (hunk 4 above)
Comment (dneto0) [reply wgsl/index.bs]: Done. PTAL.

**Pull request merged**